Fix file permissions template when checking directories and allowing stricter permissions

ggbecker opened the pull request:
#### Description:

- Update1: Fixed the template to properly assess directory permissions

#### Rationale:

pep8speaks commented:
Hello @ggbecker! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`shared/templates/file_groupowner/template.py`](https://github.com/ComplianceAsCode/content/blob/835173bb0f219ff9fbd7465cc19ab7d89416bf36/shared/templates/file_groupowner/template.py):

> [Line 16:100](https://github.com/ComplianceAsCode/content/blob/835173bb0f219ff9fbd7465cc19ab7d89416bf36/shared/templates/file_groupowner/template.py#L16): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (118 > 99 characters)

* In the file [`shared/templates/file_owner/template.py`](https://github.com/ComplianceAsCode/content/blob/835173bb0f219ff9fbd7465cc19ab7d89416bf36/shared/templates/file_owner/template.py):

> [Line 16:100](https://github.com/ComplianceAsCode/content/blob/835173bb0f219ff9fbd7465cc19ab7d89416bf36/shared/templates/file_owner/template.py#L16): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (118 > 99 characters)

* In the file [`shared/templates/file_permissions/template.py`](https://github.com/ComplianceAsCode/content/blob/835173bb0f219ff9fbd7465cc19ab7d89416bf36/shared/templates/file_permissions/template.py):

> [Line 16:100](https://github.com/ComplianceAsCode/content/blob/835173bb0f219ff9fbd7465cc19ab7d89416bf36/shared/templates/file_permissions/template.py#L16): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (135 > 99 characters)
> [Line 18:100](https://github.com/ComplianceAsCode/content/blob/835173bb0f219ff9fbd7465cc19ab7d89416bf36/shared/templates/file_permissions/template.py#L18): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (118 > 99 characters)

JAORMX commented:
/ok-to-test
/test all

ggbecker commented:
/retest

ggbecker commented:
/retest

JAORMX commented:
Well, CI seems happy

JAORMX commented:
/test all

JAORMX commented:
/test e2e-aws-ocp4-moderate

The test passed but the pod that uploads logs (provided by the CI infra) failed.

yuumasato commented:
I think the issue is actually on the scanner side, see linked PR in OpenSCAP.

JAORMX commented:
@yuumasato should we then merge https://github.com/ComplianceAsCode/content/pull/6706 and only merge that default when the scanner fix lands?

yuumasato commented:
> @yuumasato should we then merge #6706 and only merge that default when the scanner fix lands?

Either that, or we merge #6709, as it seems to provide a reasonable workaround.

JAORMX commented:
> > @yuumasato should we then merge #6706 and only merge that default when the scanner fix lands?
> 
> Either that, or we merge #6709, as it seems to provide a reasonable workaround.

@yuumasato @ggbecker whatever you guys think is best.

yuumasato commented on shared/templates/file_permissions/oval.template:
```diff
@@ -32,6 +34,9 @@
         <unix:filename operation="pattern match">{{{ FILE_REGEX }}}</unix:filename>
       {{%- else %}}
         <unix:filename xsi:nil="true" />
+      {{%- if ALLOW_STRICTER_PERMISSIONS %}}
+        <filter action="include">state_file_permissions{{{ FILEID }}}_mode_not_{{{ FILEMODE }}}</filter>
+      {{%- endif %}}
```
The use of `filter action=include` here is not wrong, but may require some explanation as to why we go the approach of "defining that the object with a certain state doesn't exist" opposed "defining the object and checking that it doesn't match a state".

yuumasato replied:
The filter should be kept until 1.3.5 is widely deployed, but there is no harm in keeping the filter indefinitely.

ggbecker replied:
I'll add a comment. I'll also try to unify this filter with the existing one that is being used for the file option (when the check is configured to look for a file)

yuumasato commented:
> @yuumasato @ggbecker whatever you guys think is best.

I'd go with this PR.
The use of `filter action="include"` enables OpenSCAP < 1.3.5 to use the file_permissions template with `allow stricter permissions`.

Revertint the template to `allow_stricter_permissions: false` won't make it work for rules who enable it.

ggbecker commented on shared/templates/file_permissions/oval.template:
```diff
@@ -20,7 +20,9 @@
       <unix:file_test check="all" check_existence="{{{ FILE_EXISTENCE }}}" comment="Testing mode of {{{ FILEPATH }}}" id="test_file_permissions{{{ FILEID }}}" version="2">
       <unix:object object_ref="object_file_permissions{{{ FILEID }}}" />
+      {{%- if not ALLOW_STRICTER_PERMISSIONS %}}
       <unix:state state_ref="state_file_permissions{{{ FILEID }}}_mode_{{{ 'not_' if ALLOW_STRICTER_PERMISSIONS }}}{{{ FILEMODE }}}" />
```
I would rather remove this condition now that we know this should  work for openscap >= 1.3.5

ggbecker commented:
/retest

ggbecker commented:
> > @yuumasato @ggbecker whatever you guys think is best.
> 
> I'd go with this PR.
> The use of `filter action="include"` enables OpenSCAP < 1.3.5 to use the file_permissions template with `allow stricter permissions`, and will work after the fix is in the wild too.
> 
> Revertint the template to `allow_stricter_permissions: false` won't make it work for rules who enable it.

I've updated the pull requests with some improvements to the code structure and applied suggestions. I've also switched to read to review, please take a look again. Commits that enabled/disabled the `allow_stricter_permissions` were stripped away from the PR.

ggbecker commented:
/retest

ggbecker commented:
@yuumasato please advise if e87eb6b is the best solution for:

https://github.com/ComplianceAsCode/content/runs/2130339562?check_suite_focus=true#step:5:141

`oval:ssg-test_file_permissions_backup_etc_group:tst:1 - No state should be referenced when check_existence has a value of 'none_exist'.`

yuumasato commented on shared/templates/file_permissions/oval.template:
```diff
@@ -1,3 +1,19 @@
+        do not match the permissions (STATEMODE). The criterion is then negated to make
+        sure there are no files/directories matched.
+    #}}
+  {{% set FILE_EXISTENCE = "at_least_one_exists" %}}
```
I think this should be `any_exist`.
If there are no files/directories with extra permissions bits set, we are fine.

ggbecker replied:
If we use `any_exist` every check would fail. For example, if no object is collected (there are no files/directories with wrong permissions), this check would pass (`any_exist` allows) and subsequentially fail because the criterion has `negate=true`. That means, we should keep the `at_least_one_exists` to cover this case.

yuumasato replied:
You are right, the `negate` confused me, :)

yuumasato commented on shared/templates/file_permissions/oval.template:
```diff
@@ -1,3 +1,19 @@
+{{%- if ALLOW_STRICTER_PERMISSIONS %}}
+    {{#
+        When allowing stricter permissions, the check looks for files/directories that
+        do not match the permissions (STATEMODE). The criterion is then negated to make
```
Text suggestion:
`looks for files/directories that have any extra permission bit set.`

yuumasato replied:
Extra suggestion: `The permissions (STATEMODE) is negated (inverted?) and becomes a mask for permissions that should not be set.`

yuumasato reviewed (commented):
> @yuumasato please advise if [e87eb6b](https://github.com/ComplianceAsCode/content/commit/e87eb6b3ba5ca0e4d3aadcc1f56975b442db01be) is the best solution for:

This looks to be most future proof and easier to update later, :)

With OpenSCAP < 1.3.5, the filter will include the objects we want (files/directories with extra permission bits set), and although the State comparison is buggy, we know all objects evaluated there will have at least one Entity trespassing the mask, thus matching and working as expected.

With OpenSCAP >= 1.3.5, the Entity matching will work correctly and we can just remove the filter. And forgetting to remove the filter will cause no harm.

yuumasato reviewed (approved):
Nice fix @ggbecker :)

LGTM, just pending results from OCP rules.

JAORMX reviewed (approved):
Thanks for taking a look at this folks! Let's wait for CI now.

JAORMX commented:
/retest

JAORMX commented:
@ggbecker yeah... the failure is unrelated. It was the cluster logging stack failing to install (yet another operator).

JAORMX commented:
/retest

JAORMX commented:
ci/prow/e2e-aws-ocp4-cis-node passed, which was the affected job. Merging.
